Update docstring for torch.linalg.lstsq

jlukehubbard opened the pull request:
Previous documentation lacked details about the handling of over- and underdetermined systems, and made incorrect mention of MAGMA.

Fixes #85021

jlukehubbard committed:
update docstring for torch.linalg.lstsq

Previous documentation lacked details about the handling of over- and underdetermined systems, and made incorrect mention of MAGMA.

jlukehubbard commented:
I wanted to preview my changes in the PR but it seems like the documentation-building workflow needs approval before it can run. Sorry to ping you but I'm not sure who would be best to ask to approve these workflows. Thanks in advance. @IvanYashchuk @lezcano

lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1015,9 +1015,8 @@
 the output has the same batch dimensions.
 
-:attr:`driver` chooses the LAPACK/MAGMA function that will be used.
+:attr:`driver` chooses the LAPACK function that will be used.
```
```suggestion
:attr:`driver` chooses the backend function that will be used.
```

lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1029,6 +1028,13 @@
 
   - `'gelsd'` (tridiagonal reduction and SVD)
   - But if you run into memory issues: `'gelss'` (full SVD).
+  
+For CUDA input, the only valid driver is `'gels'`, which assumes that :attr:`A` is full-rank. If :attr:`A` is not square,
+the following heuristic is used to choose how the system will be solved:
+  - If the linear system to be solved is overdetermined (i.e. :attr:`A` is "tall"), 
+  then `'gels'` will solve the **least squares problem** for the system as described above.
+  - If the linear system to be solved is underdetermined (i.e. :attr:`A` is "wide" and there are infinitely many solutions),
+  then `'gels'` will find the solution which minimizes :math:`\|X\|_F`, the minimum-norm solution.
```
This heuristic is trivial. Adding this does not add anything to the docs and makes them more convoluted. The previous way of stating things was just fine.

I'd be OK with moving the CUDA sentence here though.

jlukehubbard replied:
I think to solve #85021 I should explicitly mention what is done on the GPU in the underdetermined case. Maybe my most recent wording is acceptable?

jlukehubbard committed:
simplify explanation of underdetermined case

Maybe this is a better way of solving https://github.com/pytorch/pytorch/issues/85021

jlukehubbard committed:
further simplification

[skip ci] because this commit doesn't contain any changes to code, and I did not edit any runnable code in the docstring which was edited by previous commit

jlukehubbard committed:
[skip ci] final wording

lezcano reviewed (approved):
Great, thank you!

lezcano commented:
@pytorchbot merge -f "trivial doc changes"
